### ISSUE TITLE
Editable.changeFormats() should update setState

### ISSUE THREAD

**Issue by BE-Webdesign** [WordPress/gutenberg#1124]
Currently:
Around line 368 inside changeFormats()

```js
this.setState( {
	formats: merge( {}, this.state.formats, formats ),
} );
```

Should be:
```js
this.setState( ( previousState ) => ( {
	formats: merge( {}, previousState.formats, formats ),
} ) );
```

**Comment by aduth** [comment]
Can you explain the rationale on this one?

Or is this because Facebook has more recently advocated the function form as being more aligned with Fiber goals? Saddens me that it's a bit more verbose 😞

**Comment by aduth** [comment]
Noticed there's a handful more of these issues. Is there a parent issue? Should we create one?

**Comment by youknowriad** [comment]
The logic I think is because `setState` is not guaranteed to be synchronous, so we're not sure to have the state in the proper shape if we call two synchronous `setState` calls that rely on the previous state.

**Comment by aduth** [comment]
Hmm, just doesn't seem like a gotcha that's commonly encountered. At least it's not relevant in this case.

I'm not particularly opposed to it, especially now that it's become more advocated as the preferred pattern. Some questions like: Is `previousState` the most accurate naming, since it's technically the "current" state, isn't it? Or maybe even `state`, assuming it's not already been assigned in the containing scope.

Was there some previous discussion of this?

**Comment by youknowriad** [comment]
> Was there some previous discussion of this?

I don't think so. 

> Wording

Maybe `state` is a good name to match Redux's reducers

**Comment by BE-Webdesign** [comment]
Changing this actually fixes up some weird glitches I have around formatting.  I would encounter things where the format was not applying correctly and after I switched the code over it worked perfectly.

>Noticed there's a handful more of these issues. Is there a parent issue? Should we create one?

Yeah I just opened them to chronicle what I saw in a quick glance. Most of them probably will never cause an error, this one has for me.  Whether we should create a parent issue is up to you. The benefit is minimal probably, but this is a best practice and will work better with the new way fiber is going to reconcile and update.
